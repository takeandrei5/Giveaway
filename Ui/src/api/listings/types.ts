import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
