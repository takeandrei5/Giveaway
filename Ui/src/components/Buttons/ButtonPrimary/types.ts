import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
