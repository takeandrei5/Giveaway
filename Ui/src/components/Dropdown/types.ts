export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
