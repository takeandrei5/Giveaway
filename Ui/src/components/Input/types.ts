export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
