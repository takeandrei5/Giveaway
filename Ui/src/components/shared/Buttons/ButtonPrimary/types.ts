import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = ButtonProps;
