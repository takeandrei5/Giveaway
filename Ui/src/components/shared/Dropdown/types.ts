export type DropdownProps = {
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
