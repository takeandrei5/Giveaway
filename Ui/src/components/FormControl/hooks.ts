import { FieldMetaProps } from 'formik';
import { useEffect, useState } from 'react';

import { useCheckFormIsInvalidProps } from './types';

const useCheckFormIsInvalid = (meta: FieldMetaProps<unknown>): useCheckFormIsInvalidProps => {
	const [isInvalid, setIsInvalid] = useState(false);

	useEffect(() => {
		setIsInvalid(meta.touched && !!meta.error);
	}, [meta.touched, meta.error]);

	return { isInvalid };
};

export { useCheckFormIsInvalid };
