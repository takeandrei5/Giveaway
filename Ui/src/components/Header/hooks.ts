import { NextRouter, useRouter } from 'next/router';
import { useCallback } from 'react';

const useLogin = () => {
	const router: NextRouter = useRouter();

	const handleSignIn: () => Promise<boolean> = useCallback(
		async () => await router.replace('/api/auth/login'),
		[router]
	);
	const handleSignInWithReturnTo: (returnTo: string) => Promise<boolean> = useCallback(
		async (returnTo: string) => await router.replace(`/api/auth/login?returnTo=${returnTo}`),
		[router]
	);
	const handleSignOut: () => Promise<boolean> = useCallback(() => router.replace('/api/auth/logout'), [router]);

	return { handleSignIn, handleSignInWithReturnTo, handleSignOut };
};

export { useLogin };
