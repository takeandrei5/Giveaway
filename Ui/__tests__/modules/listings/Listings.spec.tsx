import { Listings } from '@modules/listings';
import { useInfiniteFetchListings } from '@modules/listings/hooks';
import { ListingsProps } from '@modules/listings/types';
import store from '@redux/store';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { SortingType } from '@utils/types';
import { getEventListeners } from 'events';

jest.mock('dateformat', () => jest.fn());

jest.mock('@modules/listings/hooks', () => ({
	useInfiniteFetchListings: jest.fn(),
}));

const wrapper = ({ children }: { children: JSX.Element }) => <Provider store={store}>{children}</Provider>;

describe('Listings', () => {
	const props: ListingsProps = {
		options: [
			{
				value: 1,
				displayValue: 'test1',
			},
			{
				value: 2,
				displayValue: 'test1',
			},
			{
				value: 3,
				displayValue: 'test1',
			},
		],
	};

	it('should match snapshot', () => {
		// Arrange
		(useInfiniteFetchListings as unknown as jest.Mock).mockImplementation(() => ({
			isLoading: false,
			totalData: [],
			nextData: {},
			sort: 'Title ASC',
			setSort: jest.fn(),
			refetchListings: jest.fn(),
		}));

		// Act
		const { container } = render(<Listings {...props} />, { wrapper });

		// Assert
		expect(container).toMatchSnapshot();
	});

	it('should execute onChangeHandler correctly', async () => {
		// Arrange
		const setSortMock = jest.fn<void, [val: SortingType]>();

		(useInfiniteFetchListings as unknown as jest.Mock).mockImplementation(() => ({
			isLoading: false,
			totalData: [],
			nextData: {},
			sort: 'Title ASC',
			setSort: setSortMock,
			refetchListings: jest.fn(),
		}));

		// Act
		const { getByTestId } = render(<Listings {...props} />, { wrapper });

		await userEvent.selectOptions(getByTestId('dropdown'), getByTestId('dropdown').querySelector('option')!.value);

		// Assert
		expect(setSortMock).toHaveBeenCalledTimes(1);
	});

	describe('data refetching on scroll', () => {
		describe('when there is no more data ', () => {
			const refetchListingsMock = jest.fn();

			it('should not trigger refetch if `useInfiniteFetchListings` returns `nextData` falsy', () => {
				// Arrange
				(useInfiniteFetchListings as unknown as jest.Mock).mockImplementation(() => ({
					isLoading: false,
					totalData: [],
					nextData: undefined,
					sort: 'Title ASC',
					setSort: jest.fn(),
					refetchListings: refetchListingsMock,
				}));

				// Act
				render(<Listings {...props} />, { wrapper });
				fireEvent.scroll(window, { target: { scrollY: 500 } });

				// Assert
				expect(refetchListingsMock).not.toHaveBeenCalled();
			});

			it('should not trigger refetch if `useInfiniteFetchListings` returns `nextData.hasNextPage` false', () => {
				// Arrange
				const setSortMock = jest.fn<void, [val: SortingType]>();
				const refetchListingsMock = jest.fn();

				(useInfiniteFetchListings as unknown as jest.Mock).mockImplementation(() => ({
					isLoading: false,
					totalData: [],
					nextData: {
						hasNextPage: false,
					},
					sort: 'Title ASC',
					setSort: setSortMock,
					refetchListings: refetchListingsMock,
				}));

				// Act
				render(<Listings {...props} />, { wrapper });
				fireEvent.scroll(window, { target: { scrollY: 500 } });

				// Assert
				expect(refetchListingsMock).not.toHaveBeenCalled();
			});
		});

		describe('when there is data ', () => {
			it('should not refetch if `useInfiniteFetchListings` returns `nextData.hasNextPage` true', () => {
				// Arrange
				const refetchListingsMock = jest.fn();

				(useInfiniteFetchListings as unknown as jest.Mock).mockImplementation(() => ({
					isLoading: false,
					totalData: [],
					nextData: {
						hasNextPage: true,
					},
					sort: 'Title ASC',
					setSort: jest.fn(),
					refetchListings: refetchListingsMock,
				}));

				// Act
				render(<Listings {...props} />, { wrapper });
				fireEvent.scroll(window, { target: { scrollY: 500 } });

				// Assert
				expect(refetchListingsMock).toHaveBeenCalledTimes(1);
			});
		});
	});
});
