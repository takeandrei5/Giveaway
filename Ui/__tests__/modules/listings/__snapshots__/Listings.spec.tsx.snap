// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listings should match snapshot 1`] = `
<div>
  <div
    class="chakra-skeleton css-r69txw"
  >
    <div
      class="css-gmuwbf"
    >
      <div
        class="css-pqsxsg"
      >
        <div
          class="chakra-stack css-n21gh5"
        >
          <span
            class="chakra-text css-16tlfhj"
            data-testid="typography"
          >
            Select from these main categories
          </span>
          <div
            class="css-1ot69hf"
          >
            <div
              class="css-8atqhb"
            >
              <div
                class="css-1o2zl33"
                data-testid="category-box"
              >
                <div
                  class="chakra-stack css-1jxjhsf"
                >
                  <div
                    class="css-120a5e1"
                  >
                    <span
                      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
                      />
                      <img
                        alt="Clothes"
                        data-nimg="responsive"
                        decoding="async"
                        draggable="false"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
                      />
                      <noscript />
                    </span>
                  </div>
                  <span
                    class="chakra-text css-1ig778p"
                    data-testid="typography"
                  >
                    Clothes
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-8atqhb"
            >
              <div
                class="css-1o2zl33"
                data-testid="category-box"
              >
                <div
                  class="chakra-stack css-1jxjhsf"
                >
                  <div
                    class="css-120a5e1"
                  >
                    <span
                      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
                      />
                      <img
                        alt="Toys"
                        data-nimg="responsive"
                        decoding="async"
                        draggable="false"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
                      />
                      <noscript />
                    </span>
                  </div>
                  <span
                    class="chakra-text css-1ig778p"
                    data-testid="typography"
                  >
                    Toys
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-8atqhb"
            >
              <div
                class="css-1o2zl33"
                data-testid="category-box"
              >
                <div
                  class="chakra-stack css-1jxjhsf"
                >
                  <div
                    class="css-120a5e1"
                  >
                    <span
                      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
                      />
                      <img
                        alt="Books"
                        data-nimg="responsive"
                        decoding="async"
                        draggable="false"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
                      />
                      <noscript />
                    </span>
                  </div>
                  <span
                    class="chakra-text css-1ig778p"
                    data-testid="typography"
                  >
                    Books
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-8atqhb"
            >
              <div
                class="css-1o2zl33"
                data-testid="category-box"
              >
                <div
                  class="chakra-stack css-1jxjhsf"
                >
                  <div
                    class="css-120a5e1"
                  >
                    <span
                      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
                      />
                      <img
                        alt="Electronics"
                        data-nimg="responsive"
                        decoding="async"
                        draggable="false"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
                      />
                      <noscript />
                    </span>
                  </div>
                  <span
                    class="chakra-text css-1ig778p"
                    data-testid="typography"
                  >
                    Electronics
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-13dls4l"
    >
      <div
        class="css-13n1q9m"
      >
        <span
          class="chakra-text css-vjlvf3"
          data-testid="typography"
        >
          Sort by: 
        </span>
        <div
          class="chakra-select__wrapper css-ts3co9"
        >
          <select
            class="chakra-select css-31afkg"
            data-testid="dropdown"
            id="sort-dropdown"
            name="sort-dropdown"
          >
            <option
              value="1"
            >
              test1
            </option>
            <option
              value="2"
            >
              test1
            </option>
            <option
              value="3"
            >
              test1
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: auto;"
      >
        <div
          class="css-ugh95v"
        />
      </div>
    </div>
  </div>
</div>
`;
